File: app/src/main/java/com/airfreshener/telegram_sms/logcat_activity.java
package com.airfreshener.telegram_sms;

import android.content.Context;
import android.os.Bundle;
import android.os.FileObserver;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.airfreshener.telegram_sms.static_class.log_func;

import org.jetbrains.annotations.NotNull;

public class logcat_activity extends AppCompatActivity {

    private Context context;
    private file_observer observer;
    private TextView logcat_textview;
    private final int line = 100;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        context = getApplicationContext();
        setContentView(R.layout.activity_logcat);
        logcat_textview = findViewById(R.id.logcat_textview);
        this.setTitle(R.string.logcat);
        logcat_textview.setText(log_func.read_log(context, line));
        observer = new file_observer(context, logcat_textview);
    }

    @Override
    public void onPause() {
        super.onPause();
        observer.stopWatching();
    }

    @Override
    public void onResume() {
        super.onResume();
        logcat_textview.setText(log_func.read_log(context, line));
        observer.startWatching();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.logcat_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        log_func.reset_log_file(context);
        return true;
    }

    private class file_observer extends FileObserver {
        private final Context context;
        private final TextView logcat;

        file_observer(@NotNull Context context, TextView logcat) {
            super(context.getFilesDir().getAbsolutePath());
            this.context = context;
            this.logcat = logcat;
        }

        @Override
        public void onEvent(int event, String path) {
            if (event == FileObserver.MODIFY && path.contains("error.log")) {
                runOnUiThread(() -> logcat.setText(log_func.read_log(context, line)));
            }
        }
    }
}

File: app/src/main/java/com/airfreshener/telegram_sms/LogcatActivity.kt
package com.airfreshener.telegram_sms

import android.os.Bundle
import android.os.FileObserver
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.airfreshener.telegram_sms.utils.LogUtils

class LogcatActivity : AppCompatActivity() {

    private var observer: LogcatFileObserver? = null
    private var logcatTextView: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_logcat)
        val logcatTextView = findViewById<TextView>(R.id.logcat_textview).apply { logcatTextView = this }
        this.setTitle(R.string.logcat)
        logcatTextView.text = LogUtils.readLog(applicationContext, LINES_COUNT)
        observer = LogcatFileObserver(applicationContext.filesDir.absolutePath, logcatTextView)
    }

    public override fun onPause() {
        super.onPause()
        observer?.stopWatching()
    }

    public override fun onResume() {
        super.onResume()
        logcatTextView?.text = LogUtils.readLog(applicationContext, LINES_COUNT)
        observer?.startWatching()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.logcat_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        LogUtils.resetLogFile(applicationContext)
        return true
    }

    private inner class LogcatFileObserver(
        path: String,
        private val logcat: TextView
    ) : FileObserver(path) {
        override fun onEvent(event: Int, path: String?) {
            if (event == MODIFY && path?.contains("error.log") == true) {
                runOnUiThread { logcat.text = LogUtils.readLog(logcat.context, LINES_COUNT) }
            }
        }
    }

    companion object {
        private const val LINES_COUNT = 100
    }
}
